src/module/components/roll-engine/engine.js

```diff
@@ -0,0 +1,486 @@
+		else return this._handleYZRoll(formData);
+	}
+
+	_validateForm(event, formData) {
+		const isEmpty = Object.values(formData).every((value) => !value);
+		const invalidArtifactField = !this.constructor.isValidArtifact(formData.artifact);
+		if (isEmpty) {
+			ui.notifications.error("No Dice Values input");
+			event.target.base.focus();
+			throw new Error("No Dice Values input");
+		}
+		if (invalidArtifactField) {
+			event.target.artifact.focus();
+			ui.notifications.error("Artifact Die input not recognized");
+			throw new Error("Artifact Die input not recognized");
+		}
+		return;
+	}
+
+	async _handleRollSpell({ base, power }) {
+		this.b = base;
+		this.damage = power;
+		const actor = FBLRollHandler.getSpeaker({
+			actor: this.options.actorId,
+			scene: this.options.sceneId,
+			token: this.options.tokenId,
+		});
+		const subtractValue = this.spell.willpower.max + 1 - this.spell.willpower.value;
+		await FBLRollHandler.modifyWillpower(actor, subtractValue, "subtract");
+		return this.executeRoll();
+	}
+
+	_handleYZRoll({ base, skill, gear, artifact, modifier }) {
+		this.b = base;
+		this.s = skill;
+		this.g = gear;
+		this.a = artifact;
+		if (modifier) this.modifyRoll(skill, modifier);
+		return this.executeRoll();
+	}
+
+	generateTermFormula(number, term, flavor = "") {
+		if (!number) return;
+		return `${number}d${term}${flavor ? `[${flavor}]` : ""}`;
+	}
+
+	parseArtifacts(string = "", artifactName = "") {
+		const artifacts = string.split(/[+, ]/).filter((term) => !!term);
+		const terms = artifacts
+			.reduce((array, artifact) => {
+				let [num, term] = artifact.split(/d/);
+				num = Number(num) || 1;
+				const existTermIndex = array.findIndex((termVal) => termVal[0] === term);
+				if (existTermIndex > -1) array[existTermIndex][1] += num;
+				else array.push([term, num]);
+				return array;
+			}, [])
+			.map(([term, num]) => this.generateTermFormula(num, term, artifactName));
+		return terms.join("+");
+	}
+
+	modifyRoll(skill, modifier) {
+		const modifiedSkill = parseInt(skill) + parseInt(modifier);
+		if (modifiedSkill < 0) {
+			this.s = 0;
+			this.n = modifiedSkill;
+		} else {
+			this.s = modifiedSkill;
+		}
+	}
+
+	getRollData() {
+		const unlimitedPush = game.settings.get("forbidden-lands", "allowUnlimitedPush");
+		// eslint-disable-next-line no-nested-ternary
+		const maxPush = unlimitedPush ? Infinity : this.options.actorType === "monster" ? "0" : 1;
+		return {
+			name: this.title,
+			maxPush: this.options.maxPush || maxPush,
+			type: this.options.type,
+		};
+	}
+
+	getRollOptions() {
+		return {
+			actorId: this.options.actorId,
+			actorType: this.options.actorType,
+			alias: this.options.alias,
+			attribute: this.base.name,
+			chance: this.spell.chance,
+			isAttack: this.isAttack,
+			damage: this.damage,
+			tokenId: this.options.tokenId,
+			sceneId: this.options.sceneId,
+			item: this.gear.name,
+			itemId: this.gear.itemId,
```

Comment:
I would like you to add this information on that line as well.
This information is important for me to be able to integrate the spell animations

```javascript
    ...
    itemId: this.gear.itemId || this.spell.item.id,
    ...
```